moneriote/moneriote.py

```diff
@@ -0,0 +1,234 @@
+import sys
+import re
+import os
+import subprocess
+import time
+from functools import partial
+from multiprocessing import Pool
+from subprocess import Popen
+from datetime import datetime
+
+from moneriote import PATH_CACHE, CONFIG
+from moneriote.dns import DnsProvider
+from moneriote.rpc import RpcNode, RpcNodeList
+from moneriote.utils import log_msg, log_err, make_json_request, banner
+
+
+if sys.version_info[0] != 3 or sys.version_info[1] < 3.5:
+    log_err("please run with python >= 3.5", fatal=True)
+
+try:
+    import requests
+except ImportError:
+    log_err("please install requests: pip install requests", fatal=True)
+
+
+class Moneriote:
+    def __init__(self, dns_provider: DnsProvider, md_address: str = '127.0.0.1', md_port: int = 18081,
+                 md_auth: str = 'not:used', md_path: str = 'monerod.exe',
+                 md_height_discovery_method: str = 'xmrchain'):
+        self.dns_provider = dns_provider
+
+        self.md_path = md_path
+        self.md_daemon_addr = md_address
+        self.md_daemon_port = md_port
+        self.md_daemon_auth = md_auth
+        if md_height_discovery_method not in ['xmrchain', 'monerod', 'compare', 'moneroblocks']:
+            log_err('bad height_discovery_method option', fatal=True)
+        self.md_height_discovery_method = md_height_discovery_method
+
+        # default Monero RPC port
+        self._m_rpc_port = 18089
+        self._blockchain_height = None
+
+        if not os.path.isfile(PATH_CACHE):
+            log_msg("Auto creating \'%s\'" % PATH_CACHE)
+            f = open(PATH_CACHE, 'a')
+            f.write('[]')
+            f.close()
+
+        self.monerod_check()
+
+    def main(self):
+        # get & set the current blockheight
+        height = self.monerod_get_height(method=self.md_height_discovery_method)
+        if not height or not isinstance(height, int):
+            log_err("Unable to fetch the current blockchain height")
+            return
+        self._blockchain_height = height
+
+        nodes = RpcNodeList()
+        nodes += RpcNodeList.cache_read(PATH_CACHE)  # from `cached_nodes.json`
+        if nodes:
+            nodes = self.scan(nodes, remove_invalid=True)
+
+        if len(nodes.nodes) <= 2:
+            peers = self.monerod_get_peers()  # from monerod
+            nodes += self.scan(peers, remove_invalid=True)
```

Comment:
There are usually 30+ 18089 opennodes online. Many of them are 24/7. So `len(nodes.nodes) <= 2` will likely make loading peer and scan never happen after first launch. It may result the peer list barely updates for adding new nodes. Maybe we could make it update at regular intervals likes once a day?

Reply:
Actually, scanning will happen every 1 hour regardless. 

The following code is from the `cached_nodes.json` read function:

https://github.com/skftn/moneriote-python/blob/2a1222774b11c0c7f47fdac4d39008e43d9a8b17/moneriote/rpc.py#L95-L99

Which invalidates cached nodes that are older than `scan_interval`, which by default is 1 hour (but you can set it to whatever)

Reply:
Sorry for didn't notice that. That's good enough!